### DISCUSSION TITLE
Is it possible to get the FSRL of a program's container from an analyser?

### DISCUSSION THREAD

**Discussion by agatti** [NationalSecurityAgency/ghidra#4465] [Q&A]
As per the subject - I have an analyser that needs to read data from a file's container that cannot be quickly brought in from its loader (indirect imports address resolution), along with the container's filesystem metadata to traverse.  While I can still serialise the filesystem tree in the loader and attach it as a property for every image file \*, I cannot know where import ordinal tables are read from in the container at loading time.

The last resort solution would be to attach the whole container's data as a program property (more than 50 megabytes at the very least), but I hope there is a cleaner way here.

*: and that brings in another question, are analysers created once per program, once per analysis request, or are just created once and reused for all programs where the analyser can be applied?  In the latter two cases, it would mean I'd have to deserialise the properties every time, which would slow analysis down to an unacceptable amount...

**Comment by ryanmkurtz** [comment]
Are you asking about how to get the `FSRL` property that's presented to you in the `Help -> About <program>` window from an Analyzer?

**Comment by agatti** [comment]
Pretty much - I suppose it can be accessed indirectly from `Program` but I wasn't able to find how.

**Comment by ryanmkurtz** [comment]
```
Options options = program.getOptions(Program.PROGRAM_INFO);
String fsrl = options.getString(ProgramMappingService.PROGRAM_SOURCE_FSRL, null);
```

**Reply by dev747368** [reply]
only thing I would add is:

```java
Options options = program.getOptions(Program.PROGRAM_INFO);
String fsrlStr = options.getString(ProgramMappingService.PROGRAM_SOURCE_FSRL, null);
FSRL fsrl = FSRL.fromString(fsrlStr);
```

**Comment by dev747368** [comment]
re: lifetime / lifecycle of Analyzers

Its not strictly defined right now.  The assumption is that the Analyzers are stateless and get a Program object with each call, but in reality, some liberties can be taken.

Currently, an Analyzer instance is created for each Program, and not re-used between Programs.  The same instance could be re-used for the same Program during the user's time in the tool if analysis is started several times.  See AutoAnalysisManager.initializeAnalyzer().

**Comment by agatti** [comment]
@ryanmkurtz, @dev747368 thanks a lot for the tips!
